### ISSUE TITLE
Proposal: Adopting Docker images as distribution/installation method

### ISSUE THREAD

**Issue by flokoe** [just-containers/s6-overlay#592]
Hi there,

I would like to propose adopting a new distribution/installation method in addition to the existing one via the `ADD` directive and making the content of the tarballs also available as Docker images so that we can use the `COPY` directive instead. This would have the following benefits:

- The version of Docker images can be easily pinned.
- Docker tags/digest can be auto-updated via Bots like Renovate/Dependabot out-of-the-box.
- `COPY` directives from images can be cached.
- The `--link` option for the `COPY` directive allows rebasing, which improves build time and caching.
-  Architecture selection could be automated via multi-arch images.

I see two possibilities: one image per tarball, or a single image containing the contents of the tarballs in corresponding subdirectories.

- **single image:**
  - Version compatibility between files.
  - Single image to update.
  - The different tarballs are kind of "hidden" (not directly mentioned in the image name).
  - Single layer.
- **image per tarball:**
  - The tarballs are easier to differentiate.
  - Multiple layers.
  - Users have to ensure that each image has the same version, could lead to version mismatches.

It could look something like this:

```Dockerfile
# Singe image with digest, copying multiple "tarballs" into new image.
# A multi-arch image would automatically select the correct architecture.
FROM alpine:latest

COPY --link \
  --from=just-containers/s6-overlay:v3.2.0.0@sha256:17d70c7e8e374f772c0d0539f2e794b15701a50c48cc52cebea20d96e017c95f \
  /s6-overlay-noarch /s6-overlay-arch /


# Different images per tarball.
# A multi-arch image would automatically select the correct architecture.
FROM alpine:latest

COPY --link \
  --from=just-containers/s6-overlay-noarch:v3.2.0.0@sha256:933e09abf7298e4af67d18a54391b65c207e2181e143b5489e064b3078f37aed \
  / /

COPY --link \
  --from=just-containers/s6-overlay-arch:v3.2.0.0@sha256:4d6c5e083fb05bb2700740cc0a2f26e1307afac63da3ccf3d81efa0a0e8375f2 \
  / /
```

File permissions are preserved and directories are merged. Any conflicts are resolved in favor of the content being added.

I would love to hear your thoughts on this. If you are interested, I could submit a PR with the necessary Dockerfiles and GitHub Actions workflows to automate the builds.

**Comment by skarnet** [comment]
Sorry, I have exactly zero available bandwidth at the moment, I'll get back to you later. @jprjr, if you're around, can you please take a look and say if it's reasonable and doesn't require too much work?

**Comment by flokoe** [comment]
> Sorry, I have exactly zero available bandwidth at the moment, I'll get back to you later. @jprjr, if you're around, can you please take a look and say if it's reasonable and doesn't require too much work?

No worries. Like I said, I am happy to provide all the necessary Dockerfiles, workflows, docs etc. if you are interested.

**Comment by crazy-max** [comment]
Seems similar to this proposal https://github.com/just-containers/s6-overlay/issues/386#issuecomment-1053055397

**Comment by skarnet** [comment]
@flokoe Okay, sorry for the delay. So, I have a few comments about the benefits of that approach:

- Versions can be pinned: see, that's a direct *drawback* for me. Pinning is a quick and dirty fix that should be temporary but most often overstays its welcome and accumulates tech debt. I don't want people to feel like they have to pin a version of s6-overlay. I'd rather make it possible for them to trust that upgrades won't break their workflow. So, ease of pinning isn't an argument for me, on the contrary.
- Docker tags/digests can be updated by bots: I have no idea what it means 😅 Can you please explain like I'm five?
- COPY allows caching etc: yeah, sure, but the impact of this should be pretty small. Are people spending their days building images, rather than running them?
- Architecture selection: so the thing is, I absolutely do not want to be tied to the Docker way of describing architectures, which is *inferior to* the one we have at the moment and that allows us to target architectures that are not explicitly supported by Docker. How would the automated architecture selection play with the current system?

More generally, I said initially that I didn't want to be in charge of the s6-overlay project: I maintain it because it's useful to people, and because it's currently the most popular use of s6. I am okay with the *development* side of things, and we had an agreement with @jprjr that he would handle the *ops* side of things. But it doesn't seem like he wants to be involved any more; so if it's going to be just me, I'm going to touch the ops side as little as possible. Containers are something I'm involved with because it's where my software is used; but I don't want them to consume my time, so if it's not broken, I'm not going to fix it.

So, yeah, if you're willing to provide a proof of concept with Dockerfiles and GitHub actions that I could play with to see how it would go, sure, why not; but expect to have to provide a lot of patient explanations, and opinions and point of views that are probably rather different from your usual container software provider. Sorry for being so conservative and reluctant about this; but managing my priorities is absolutely necessary.

**Comment by skarnet** [comment]
@crazy-max It doesn't sound similar to your project; basically, what you're doing is rewriting the build system in the Dockerfile language, which I don't see the benefit of at all. Whereas if I understand correctly, @flokoe seems to want to keep the existing build, just provide different end outputs via Github actions voodoo, to integrate with various user workflows.

**Comment by socheatsok78** [comment]
I would agree with @flokoe. But only container image is all that we need, e.g. `just-containers/s6-overlay`. 

We can distribute the tarballs using the tags. Example:
- `s6-overlay-noarch.tar.xz` and `s6-overlay-${arch}.tar.xz` as `v3.2.0.0`
- `s6-overlay-symlinks-noarch.tar.xz` and `s6-overlay-symlinks-arch.tar.xz` as `v3.2.0.0-symlinks`
- `syslogd-overlay-noarch.tar.xz` as `v3.2.0.0-syslogd`

All that we need after that for `Dockerfile`:

```Dockerfile
ARG S6_OVERLAY_VERSION=v3.2.0.0
FROM containers/s6-overlay:${S6_OVERLAY_VERSION} AS s6-overlay
#FROM containers/s6-overlay:${S6_OVERLAY_VERSION}-symlinks AS s6-overlay-symlinks
#FROM containers/s6-overlay:${S6_OVERLAY_VERSION}-syslogd AS s6-overlay-syslogd

FROM alpine:latest
COPY --link --from=s6-overlay / /
#COPY --link --from=s6-overlay-symlinks / /
#COPY --link --from=s6-overlay-syslogd / /
ENTRYPOINT [ "/init" ]
```

This is what I've made and use personally [socheatsok78/s6-overlay-distribution](https://github.com/socheatsok78/s6-overlay-distribution), previously [socheatsok78/s6-overlay-installer](https://github.com/socheatsok78/s6-overlay-installer]) but I found that it inconvenience.

**Comment by tcely** [comment]
I'm interested in using something like this.

If either @flokoe or @socheatsok78 want to maintain a distribution I could `COPY` from instead that would simply my current`Dockerfile` rather a lot. So much the better if it gets automated as part of a new release.

**Comment by jprjr** [comment]
Hey, I just wanted to chime in - it's honestly probably about time I get officially retire from the project. I haven't been very responsive, and that's really not great of me - I sincerely apologize for that.

When we started all of this - I was using Docker pretty heavily at my job, making containers for services, orchestration, the works. I got to take a lot of what I was doing and learning at work, then apply it to this.

Since then, I've moved, changed jobs a few times - I really don't use Docker anymore at my job (my employer does, but my particular role doesn't). It's doubtful I'll be really using a lot of Docker in the future.

It's not that I don't *want* to be involved, it's just gotten really hard lately. Career-wise, it seems like I'm moving out of this particular field and into others, and that's really hampering my ability to participate.

Again, I really do apologize for being non-responsive. I think at some point it seemed like things didn't need a lot of my involvement and I kind of shut off. I was just going through my GitHub inbox tonight and realized, I was actually specifically mentioned in this issue! Months ago!

So, thought I should say hey, I sincerely enjoyed helping make this, I think we created something great. I notice the s6-overlay pop up all the time in various images. It's something to be proud of. I'm just not in the same position I was when we started all of this... man has it really been almost ten years? Wow! So, it's time for me to step down.

**Comment by skarnet** [comment]
@jprjr It's all good, thanks for confirming what I already thought. It's incredible that a silly idea from three random dudes went so far and had so much success - and you, with your two blog posts, are the main reason for s6-overlay's popularity, so thank you so much for your involvement so far. It's been great, and I wish you a very happy future (and I still follow you on the fediverse 😁)

What this means for the project is that, since Gorka apparently fell off the face of the Earth, I'm now alone maintaining it - which I *explicitly* did not want at start; but since s6 is my baby and s6-overlay is 1. a good piece of software, genuinely useful for people, and 2. the place where s6 is the most used by far, I will keep maintaining it for the foreseeable future, so people should not worry about that. Except for bus factor reasons, but it's likely that society as we know it will collapse before I retire or die, so we're good.

However, since I also don't use Docker in my day-to-day (so far), except to test s6-overlay, reproduce reported bugs, etc., and have zero knowledge of (and little desire to learn) GitHub Actions, it means that the ops side of s6-overlay will now move at an even more glacial pace. I will only touch things that are obviously broken. This is bad news for this proposal as well as a few other suggestions; the issues will remain open because there's no good reason to close it, but don't expect much progress on them for now.

I will look for someone to help with the ops side.

**Comment by mustanggb** [comment]
This would be wonderful.

I suggest that the single image approach would be the preferable way to go, as it makes things the most straightforward and easy to use, which is kind of the goal.

This would also resolve the architecture selection requests that keep popping up: #617, #561, #512, etc.

It seems like there is enough want and willingness to get this in.

So next steps would be for someone (probably @flokoe) to do the proof of concept.

**Comment by socheatsok78** [comment]
This is what I've been maintaining and using personally. Here an example:

```Dockerfile
# example/Dockerfile

ARG S6_VERSION=v3.2.0.0
FROM socheatsok78/s6-overlay:${S6_VERSION} AS s6-overlay
# FROM socheatsok78/s6-overlay:${S6_VERSION}-symlinks AS s6-overlay-symlinks  # <- uncomment this overlay symlinks

# This is your final image
FROM alpine:latest

# Overlay the s6-overlay files to the root filesystem
# The order of this is up to you, but generally, you want to copy the main s6-overlay first,
# followed by symlinks and syslogd if needed.
COPY --link --from=s6-overlay / /
# COPY --link --from=s6-overlay-symlinks / /  # <- uncomment this overlay symlinks
ENTRYPOINT [ "/init" ]
CMD [ "/bin/sh" ]

# Add any additional packages or configurations you need here
```

The `socheatsok78/s6-overlay` uses the Docker Buildx to build multi-platform container images using `FROM scratch` as the base image (`FROM scratch` is a no-op in the `Dockerfile`). This way you can overlay the `socheatsok78/s6-overlay` on top of any base image the user want to be running as seen in the example above.

Instead of having to maintain a list of distro-based image like @WhyNotHugo mentioned, which take a lot of efforts (no offence), using `FROM scratch` to distribute `s6-overlay` make thing a lot simpler and it is the same as using the tarballs in the GitHub releases, except we don't have to think about which archive to pull and how to extract the archive in the final image. We can just overlay it on top of any existing distro you wants.

Following platforms for this image are available:

```bash
$ docker run --rm mplatform/mquery socheatsok78/s6-overlay:v3.2.0.0

# Image: socheatsok78/s6-overlay:v3.2.0.0
#  * Manifest List: Yes (Image type: application/vnd.oci.image.index.v1+json)
#  * Supported platforms:
#    - linux/amd64
#    - linux/arm64
#    - linux/riscv64
#    - linux/s390x
#    - linux/386
#    - linux/arm/v7
#    - linux/arm/v6
```

@skarnet what is your thought on this? Or maybe @crazy-max might have any suggestions since you works at @docker.

You can take a look at the implementation here [socheatsok78/docker-s6-overlay](https://github.com/socheatsok78/docker-s6-overlay). There are many ways to integrate this to the repo and CI/CD pipeline. Since CI/CD need to build multi-platform image we can still leverage Docker Buildx but the changes is gonna be substantial.

**Comment by tcely** [comment]
That looks really good to me.

There is only one thing that my current process does that yours doesn't; checking the downloaded files.

If you added download and verification of the hash for every file, then I would be confident to switch to something like what you are using.

**Comment by socheatsok78** [comment]
> There is only one thing that my current process does that yours doesn't; checking the downloaded files.

Agreed, since It mainly for personal uses. I simply didn't bother add the verification check. But adding one is not that difficult.

**Comment by tcely** [comment]
When I did this work, I found that the `sha256sum` from `busybox` wasn't behaving as I liked.

I'd suggest:
- install `cmd:sha256sum` with `apk`
- use `sha256sum --check --warn --strict`
- move any files that pass verification into a different directory

This way you will not be able to use files that were downloaded but failed to be verified.

I think I used something like `/downloaded` and `/verified` then only extracted from paths beginning with `/verified/`, but it's been a while since I looked at this stuff.

**Comment by socheatsok78** [comment]
Apology for late to response!

@skarnet thank you for the compliments, I hope that this containerization would bring a more streamline process of integrating s6-overlay since it primary use is within a container anyway. 

And to avoid robbing the users from what they already have implemented using the tarballs, We could create an official repo for `docker-library`, `official-images` or I can transfer the ownership of `socheatsok78/docker-s6-overlay` to `just-containers`. 

Integrating it directly in this repo would make it difficult to maintain for you.

I might be able to offer some helping hand from time-to-time, specially like you said regarding (answer user requests).
